contribute/design/1.x/replica_scaleup/dataplane/20190910-replica-scaleup.md

```diff
@@ -0,0 +1,278 @@
+* [Graduation Criteria](#graduation-criteria)
+* [Implementation History](#implementation-history)
+* [Drawbacks](#drawbacks)
+* [Alternatives](#alternatives)
+* [Infrastructure Needed](#infrastructure-needed)
+
+## Summary
+
+This proposal includes design for cStor data plane to allow adding new replicas
+to it so that replication factor of the volume will be increased.
+
+## Motivation
+
+There are cases arising where OpenEBS user lost multiple copies of data, and,
+working with available single copy.
+This proposal is to enable him to add more replicas to it again.
+
+### Goals
+
+- Increase ReplicationFactor of volume by adding replicas
+- Identify replicas which are allowed to connect to target
+
+### Non-Goals
+
+- Scaling down replicas
+- Add replicas in GitOps model
+- Workflow to replace non-existing replica
+- operator that detects the need to increase ReplicationFactor
+- Allow adding replicas to replace particular replica (volume distribution case)
+
+## Proposal
+
+### User Stories
+
+#### Add new replica
+As an OpenEBS user, I should be able to add replicas to my volume.
+
+#### Replace non-existing replica
+As an OpenEBS user, I should be able to replace non-existing replica with new
+replica.
+
+### Implementation Details/Notes/Constraints
+
+Currently, cstor-istgt reads replication related configuration, i.e.,
+ReplicationFactor (RF) and ConsistencyFactor (CF), from istgt.conf file.
+cstor-volume-mgmt is the container that fills the RF and CF details into
+istgt.conf file by reading from CStorVolume CR.
+
+Another property that cstor-istgt uses is 'quorum'. This property is set at
+replica, and, replica sends this to istgt during handshake phase.
+This property can take values of 'on' and 'off'.
+'on' means that data written to this replica is available for reads.
+'off' means that data written to this replica is lost.
+cstor-pool-mgmt is the container that sets quorum property at replica.
+It sets property as 'on' if replica is created first time related to that CVR.
+Otherwise, it sets to 'off'.
+
+#### Current Implementation -- Config
+- cstor-volume-mgmt reads configuration from CStorVolume CR.
+  - spec.replicationFactor
+  - spec.consistencyFactor
+- cstor-volume-mgmt updates the above info into istgt.conf file as
+  - ReplicationFactor
+  - ConsistencyFactor
+- Based on the status of CVR and availability of replica, cstor-pool-mgmt
+sets quorum property while creating replica.
+
+#### Current Implementation
+Above mentioned properties are used as below at istgt:
+- Allow replicas (upto max of 5) to connect to target until RF number of
+healthy replicas are available.
+- Atleast CF in-quorum replicas are required to perform rebuilding and start IOs
+- Rebuilding will be triggered and replicas become healthy. As RF number of
+replicas are healthy, other connected replicas, if any, gets disconnected.
+
+#### Shortcomings With Current Implementation
+- Due to misconfiguration from user, if old replica connects back instead of
```

Comment:
Example will help here. Is this the case where:
- RF=3, 
- One replica is healthy, q=on
- Two replicas went down and got added back as q=off -- rebuild started and q=on was set. 
- The node configuration changed and the two replicas came back with older disk with older data, q will remain as q=on, but data in it is old.

Reply:
yes.. this is the case I want to mention. Will provide the example.